File: worlds/network/cpp/thrift_agent_client.cpp
#include <iostream>

#include <thrift/protocol/TBinaryProtocol.h>
#include <thrift/transport/TSocket.h>
#include <thrift/transport/TTransportUtils.h>

#include "thrift_agent_client.h"

using namespace std;
using namespace apache::thrift;
using namespace apache::thrift::protocol;
using namespace apache::thrift::transport;

using namespace env;

AgentClient* setup_client() {
    std::shared_ptr<TTransport> socket(new TSocket("localhost", 9090));
    std::shared_ptr<TTransport> transport(new TBufferedTransport(socket));
    std::shared_ptr<TProtocol> protocol(new TBinaryProtocol(transport));
    AgentClient* client = new AgentClient(protocol);
    transport->open();
    return client;
}

int test() {

    AgentClient* agent = setup_client();

    Shape shape;
    shape.shape = {1};
    SpaceSpec actionSpec;
    actionSpec.shape = shape;
    Space actionSpace = {
	{"move", actionSpec}
    };

    SpaceSpec observationSpec;
    observationSpec.shape = shape;
    Space observationSpace = {
	{"move", observationSpec}
    };

    map<string, string> ret;
    agent->init(ret, actionSpace, observationSpace);
    cout << "init returned: " << endl;
    for (auto const& p : ret) {
	cout << p.first << ": " << p.second << endl;
    }

    ETensor et;
    et.shape = shape;
    et.values = {1, 2, 3};

    Observations observations = {
	{"world", et}
    };

    Actions actions;

    agent->step(actions, observations, "debug_string");

    for (auto const& p: actions) {
	cout << "action is: " << p.first << ": " << p.second << endl;
    }

    return 0;
}

/*
int main() {
    return test();
}
*/

File: models/integrated/simple_supervised_networked/thrift_agent_client.py
import network.genpy.env.ttypes
from thrift.transport import TSocket
from thrift.transport import TTransport
from thrift.protocol import TBinaryProtocol

from typing import TypedDict
from network.genpy.env.Agent import Client #see worlds\setup.py
from network.genpy.env.ttypes import ETensor, Shape, SpaceSpec #see worlds\setup.py

import random
import time

# TODO: these should probably be passed in as parameters to setup
# but for now we just use them as defaults
# or we could even define them as constants in thrift
HOST = "127.0.0.1"
PORT = '9090'


def setup_client():
    transport = TSocket.TSocket(HOST, PORT)
    transport = TTransport.TBufferedTransport(transport)
    protocol = TBinaryProtocol.TBinaryProtocol(transport)
    client = Client(protocol)
    # TODO: could make this an enterable thing so can be used like with(client)
    # which takes care of closing it also
    transport.open()
    return client


if __name__ == '__main__':

    agent = setup_client()
    actionSpaceSpec = SpaceSpec(shape=Shape(shape=[1]))
    actionSpace = {"move": actionSpaceSpec}
    observationSpaceSpec = SpaceSpec(shape=Shape(shape=[1]))
    observationSpace = {"world": observationSpaceSpec}
    res = agent.init(actionSpace, observationSpace)
    print(f'init returned: {res}')
    allpairs = []
    import torch
    all_vectors = torch.zeros(size=(30,25))
    for i in range(30):
        vals = [0.0 if random.random() < 0.8 else 1.0 for _ in range(25)]
        observation = ETensor(Shape(shape=[5, 5]), values=vals)
        observations = {"Input": observation, "Output": observation}
        all_vectors[i] = torch.Tensor(vals)
        allpairs.append(observations)
    while True:
        for i in range(30):
            print(allpairs[i])
            action:TypedDict= agent.step(allpairs[i], "debug_string")

            vector_val:torch.Tensor = torch.Tensor(action["Output"].vector.values).float()
            closest_index = torch.argmin((all_vectors - vector_val).abs().sum(dim=1).flatten())

            print(i, closest_index)
            print(f'action is: {action}')
